network-config-analyzer/CalicoPolicyYamlParser.py

```diff
@@ -45,46 +45,162 @@ def _parse_selector_expr(self, expr, origin_map, namespace=None, is_namespace_se
+            return all_peers & self.peer_container.get_peers_with_label(key, [substr],  action)
 
-        # TODO: support expressions like "key contains s", "key starts with s", "k ends with s", "global()"
         self.syntax_error('Invalid expression', origin_map)
         return None
 
+    def _strip_selector(self, expression):
+        """
+        removing brackets and spaces
+        :param str expression: The selector expression to parse
+        :return: striped expression
+        :rtype: str
+        """
+        expression = expression.strip()
+        if expression.count('(') != expression.count(')'):
+            self.syntax_error('number of brackets are not even: ' + expression)
+        while expression.startswith('(') and expression.endswith(')'):
+            n_brackets = 0
+            do_not_remove = False
+            for c in expression[0:len(expression)-1]:
+                if c == '(':
+                    n_brackets += 1
+                if c == ')':
+                    n_brackets -= 1
+                if n_brackets == 0:
+                    do_not_remove = True
+            if do_not_remove:
+                break
+            expression = expression[1:len(expression) - 1]
+            expression = expression.strip()
+        return expression
+
+    @staticmethod
+    def _split_selector(expression, operator):
+        """
+        split the expression by an operator, considering brackets. i.e. "(a || b) || c" is being split to ["(a || b)", "c"]
+        :param str expression: The selector expression to parse
+        :param str operator: The operator to split with
+        :return: a list of expressions
+        :rtype: list of str
+        """
+        expressions = expression.split(operator)
+        current_expr = ""
+        splitted_expr = []
+        for expr in expressions:
+            current_expr += expr
+            if current_expr.count('(') == current_expr.count(')'):
+                splitted_expr.append(current_expr)
+                current_expr = ""
+            else:
+                current_expr += operator
+        return splitted_expr
+
+    def _recursive_parse_label_selector(self, label_selector, origin_map, namespace, namespace_selector):
+        """
+        Recursive Parse a label selector expression appearing in selector/notSelector/namespaceSelector parts of the EntityRule
+        :param str label_selector: The selector expression to parse
+        :param dict origin_map: The EntityRule object (for reporting errors)
+        :param K8sNamespace namespace: Restrict pods to the given namespace
+        :param bool namespace_selector: True if this is a namespaceSelector
+        :return: the set of peers selected by the selector expression
+        :rtype: PeerSet
+        """
+
+        include_globals = not namespace_selector or 'global()' in label_selector
+        label_selector = self._strip_selector(label_selector)
+        # handling '||' :
```

Comment:
Maybe just add a comment here: "Order of handling operators is the reversed order of precedence"